src/cls/rbd/cls_rbd.cc

```diff
@@ -6622,6 +6622,41 @@ int trash_get(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   return r;
 }
 
+int trash_state_set(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
```

Comment:
Nit: document the in/out params encoded in the bufferlist

Reply:
I am re-adding this comment since it's still valid and was hidden by a code change:
Note that there is still a (smaller) race condition where you can start to restore and then delete the image from the trash. I guess to fully solve it would require multiple states (moving, removing, restoring) and the trash_state_update method should take an expected-state argument (or just require the state to be normal or match).

Reply:
@dillaman  Thank you. To be honest, I have no idea how to fully solve it. Closed this PR and will re-open when it's ready for a review.

Reply:
@Songweibin Sorry -- I wasn't trying to be difficult. You can either solve it by passing a third parameter to ```trash_state_set``` (which is the expected state) or have ```trash_state_set``` just check the current state to see if it's set to ```NORMAL``` or if it matches the input parameter state. Either way, if it doesn't match, just return ```-ESTALE``` or something.

Reply:
@dillaman  Thank you very much, I will try :-)